lib/include/jxl/cms_interface.h

```diff
@@ -0,0 +1,210 @@
+/* Copyright (c) the JPEG XL Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** @file cms_interface.h
+ * @brief Interface to allow the injection of different color management systems
+ * (also called color management modules, or CMMs) in JPEG XL.
+ *
+ * A CMS is needed by the JPEG XL encoder and decoder to perform colorspace
+ * conversions. This defines an interface that can be implemented for different
+ * CMSes and then passed to the library.
+ */
+
+#ifndef JXL_CMS_INTERFACE_H_
+#define JXL_CMS_INTERFACE_H_
+
+#include "jxl/color_encoding.h"
+#include "jxl/memory_manager.h"
+
+#if defined(__cplusplus) || defined(c_plusplus)
+extern "C" {
+#endif
+
+/** Represents an input or output colorspace to a color transform, as a
+ * serialized ICC profile. */
+typedef struct {
+  /** The serialized ICC profile. */
+  struct {
+    const uint8_t* data;
+    size_t size;
+  } icc;
+
+  /** Structured representation of the colorspace, if applicable. */
```

Comment:
How can one know whether color_encoding, icc, or both are being used? Should there be a flag for one, or for each?

Reply:
I imagine that for an external CMS, the convenient way to do things is to always use icc.

For the internal partial CMS that we have in libjxl (which is needed to do decoding in case of nontrivial frame blending), only color_encoding is implemented.

What is going to implement this C interface, and will it be acceptable/possible to have an implementation that is only partial, e.g. that only works with color_encoding but not with icc profiles?

Reply:
AFAIK, the user will implement a function that receive this JxlColorProfile, right (two instances)? So the user is the one who is able to use the color_encoding field if they want to. But how can they know that they can use it, if there's no indication whether it's unused, or filled in?

The JxlColorProfile could contain "unknown" but still have valid other values, in theory, the jxl file format allows this

Reply:
It should be documented what is guaranteed to be present, what is maybe there, and what takes precedence in case of conflicts. Maybe we also need a boolean indicating whether it is safe to use the struct colorspace instead of the ICC profile.

In the CMYK case, the spec requires that to be done always with an explicit CMYK ICC profile, a kBlack extra channel, and ColorSpace in the image header set to RGB (since we don't have an enum value there for CMYK, only for RGB, Gray, XYB and UNKNOWN). This is maybe a bit confusing and it might be clearer to add a JXL_COLOR_SPACE_CMYK enum value in the api, which gets returned in case there is a kBlack channel. Not that the struct colorspace can ever be used in the CMYK case, so maybe it doesn't really matter.